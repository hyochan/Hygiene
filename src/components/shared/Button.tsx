import {
  ActivityIndicator,
  TextStyle,
  TouchableOpacity,
  TouchableOpacityProps,
  ViewStyle,
} from 'react-native';

import React from 'react';
import styled from 'styled-components/native';

const StyledButton = styled.View`
  align-self: center;
  /* width: 320px; */
  height: 52px;
  border-color: blue;
  align-items: center;
  justify-content: center;
  flex-direction: row;
`;

const StyledDisabled = styled(StyledButton)`
  background-color: #cccccc;
  border-color: rgb(200,200,200);
`;

const StyledText = styled.Text`
  font-size: 16px;
  color: #069ccd;
`;

const StyledDisabledText = styled(StyledText)`
  color: #969696;
`;

interface Props {
  testID?: string;
  isLoading?: boolean;
  isDisabled?: boolean;
  onPress?: () => void;
  containerStyle?: ViewStyle;
  style?: ViewStyle;
  disabledStyle?: ViewStyle;
  textStyle?: TextStyle;
  disabledTextStyle?: TextStyle;
  leftElement?: React.ReactElement;
  rightElement?: React.ReactElement;
  indicatorColor?: string;
  activeOpacity?: number;
  text?: string;
  touchableOpacityProps?: TouchableOpacityProps;
}

function Button(props: Props): React.ReactElement {
  const {
    isDisabled,
    testID,
    disabledStyle,
    containerStyle,
    isLoading,
    style,
    textStyle,
    disabledTextStyle,
    indicatorColor,
    text,
    leftElement,
    rightElement,
    activeOpacity,
    onPress,
    touchableOpacityProps,
  } = props;

  if (isDisabled) {
    return (
      <StyledDisabled
        testID={testID}
        style={[
          containerStyle,
          disabledStyle,
        ]}
      >
        <StyledDisabledText
          style={[textStyle, disabledTextStyle]}
        >{text}</StyledDisabledText>
      </StyledDisabled>
    );
  }

  if (isLoading) {
    return (
      <StyledButton testID={testID} style={[
        containerStyle,
        style,
      ]}>
        <ActivityIndicator
          size="small"
          color={indicatorColor}
        />
      </StyledButton>
    );
  }

  return (
    <TouchableOpacity
      testID={testID}
      activeOpacity={activeOpacity}
      onPress={onPress}
      style={containerStyle}
      delayPressIn={30}
      {...touchableOpacityProps}
    >
      <StyledButton style={style}>
        {leftElement || null}
        <StyledText style={textStyle}>{text}</StyledText>
        {rightElement || null}
      </StyledButton>
    </TouchableOpacity>
  );
}

export default Button;
