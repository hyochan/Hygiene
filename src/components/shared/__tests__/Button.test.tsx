import * as React from 'react';

import Button from '../Button';
import { Text } from 'react-native';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

// eslint-disable-next-line
const component = (props?: any): React.ReactElement => {
  return <Button {...props} />;
};

describe('[Button]', () => {
  let rendered: renderer.ReactTestRenderer;
  let root: renderer.ReactTestInstance;

  it('should render without crashing', () => {
    rendered = renderer.create(component());
    expect(rendered.toJSON()).toMatchSnapshot();
    expect(rendered.toJSON()).toBeTruthy();
  });

  it('should render loading status', () => {
    rendered = renderer.create(component({ isLoading: true }));
    root = rendered.root;

    const texts = root.findAllByType(Text);

    expect(texts).toHaveLength(0);
  });

  it('should render disabled status', () => {
    rendered = renderer.create(component({ isDisabled: true }));
    root = rendered.root;

    const texts = root.findAllByType(Text);

    expect(texts).toHaveLength(1);
  });

  it('should render left and right elements', () => {
    rendered = renderer.create(component({ leftElement: <Text/>, rightElement: <Text/> }));
    root = rendered.root;

    const texts = root.findAllByType(Text);

    expect(texts).toHaveLength(3);
  });

  describe('[Button] Interaction', () => {
    let cnt = 1;

    it('simulate onPress', () => {
      rendered = renderer.create(
        component({
          onClick: () => cnt++,
        }),
      );

      root = rendered.root;

      root.findByType(Button).props.onClick();
      expect(cnt).toBe(2);
    });
  });
});
