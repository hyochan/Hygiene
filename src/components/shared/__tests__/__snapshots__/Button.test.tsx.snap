// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Button] should render without crashing 1`] = `
<View
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "borderColor": "blue",
          "flexDirection": "row",
          "height": 52,
          "justifyContent": "center",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#069ccd",
            "fontSize": 16,
          },
        ]
      }
    />
  </View>
</View>
`;
